### PR TITLE
InnoDB Support - initial version

### PR DESCRIPTION
**Pull request by olegsv**
1. Added two new commands to SQL upgrade:  

- #IfTableEngine - for tables that can not be converted by ALTER TABLE ... ENGINE=INNODB
- #IfInnoDBMigrationNeeded - to process all the other tables.

 Most of the tables can be converted using simple ALTER TABLE

2. For tables: ar_activity, claims, procedures_order_code and procedure_answers , where PRIMARY KEY is multiple columns index with AUTO_INCREMENT column - refactored the code to emulate INSERT functionality using transactions.

3. The migration process is LONG and it will be LONGER for large tables.

### CODE REVIEW

**Comment by yehster** [comment]
The modifications for multi-column key/auto insert needs to be tested when there is concurrency, and for verification that the transnational semantics work properly.  It's possible that there might also need to be an explicit lock on the table to ensure that the sequence_no is generated properly.

**Comment by yehster** [comment]
I'm also not sure what happens if the table is still MyIASM, since transaction's don't really do anything in that case.

**Comment by olegsv** [comment]
To be absolutely sure we can add FOR UPDATE . 

```
BEGIN TRANSACTION;
SELECT IFNULL(MAX(sequence_no),0) + 1 AS increment FROM ar_activity WHERE pid=1 AND encounter=1 FOR UPDATE
```
To my understanding , as long as the transaction is not finalized, MySQL will put all SELECT ... WHERE pid=1 AND encounter=1. ... FOR UPDATE queries on hold.

Also all queries with the same pid but with different encounter will be put on hold.

http://dev.mysql.com/doc/refman/5.7/en/innodb-locking.html#innodb-shared-exclusive-locks

**Comment by yehster** [comment]
The best approach might be to make determining the next sequence number a subquery of the insert statement rather than using two separate statements.  This way, it's still a single atomic operation even if the table is still MyIASM.  This isn't the exact syntax to work, but it's close enough to give the general idea.

					 sqlStatement("insert into ar_activity set "    .
 					"pid = '"       . trim(formData("HiddenPId$CountRow"   )) .
 					"', encounter = '"     . trim(formData("HiddenEncounter$CountRow"   ))  .
					"', sequence_no = '"     . SELECT IFNULL(MAX(sequence_no),0) + 1 AS increment FROM ar_activity WHERE pid = ? AND encounter = ?", (trim(formData("HiddenPId$CountRow"   )), trim(formData("HiddenEncounter$CountRow"   )  .
                                         "', code_type = '"      . trim(formData("HiddenCodetype$CountRow"   ))  .
 					"', code = '"      . trim(formData("HiddenCode$CountRow"   ))  .
 					"', modifier = '"      . trim(formData("HiddenModifier$CountRow"   ))  .

 					"', adj_amount = '"    . 0 .
 					"', account_code = '" . "$AccountCode"  .

File: interface/forms/vitalsM/table.sql [hunk 1]
```diff
@@ -21,4 +21,4 @@ CREATE TABLE IF NOT EXISTS `form_vitalsM` (
 `head_circ`         FLOAT(5,2)      default 0,
 `oxygen_saturation` FLOAT(5,2)      default 0,
 PRIMARY KEY (id)
-) ENGINE=MyISAM;
+) ENGINE=InnoDB;
```
Comment (bradymiller) [inline interface/forms/vitalsM/table.sql]: Just to be complete, would also change MyISAM to InnoDB in contrib/forms

Comment (bradymiller) [inline interface/forms/vitalsM/table.sql]: Here are the files that would need to be changed:
contrib/forms/hp_tje_primary/table.sql:) ENGINE=MyISAM AUTO_INCREMENT=1 ;
contrib/forms/hp_tje_primary/table.sql:) ENGINE=MyISAM AUTO_INCREMENT=1 ;
contrib/forms/hp_tje_primary/table.sql:) ENGINE=MyISAM AUTO_INCREMENT=1 ;
contrib/forms/hp_tje_primary/table.sql:) ENGINE=MyISAM AUTO_INCREMENT=1 ;
contrib/forms/football_injury_audit/table.sql:) ENGINE=MyISAM;
contrib/forms/sports_fitness/table.sql:) ENGINE=MyISAM;
contrib/forms/contacts/table.sql:) ENGINE=MyISAM;
contrib/forms/vision/table.sql:) ENGINE=MyISAM;
contrib/forms/scanned_notes/table.sql:) ENGINE=MyISAM;
contrib/forms/assessment_intake/table.sql:) ENGINE=MyISAM;
contrib/forms/clinical_notes/table.sql:) ENGINE=MyISAM;
contrib/forms/hist_exam_plan/table.sql:) ENGINE=MyISAM;
contrib/forms/prosthesis/table.sql:) ENGINE=MyISAM;
contrib/forms/ippf_srh/table.sql:) ENGINE=MyISAM;
contrib/forms/body_composition/table.sql:) ENGINE=MyISAM;
contrib/forms/reviewofs/table.sql:) ENGINE=MyISAM;
contrib/forms/ped_pain/table.sql:) ENGINE=MyISAM;
contrib/forms/soccer_injury/table.sql:) ENGINE=MyISAM;
contrib/forms/strength_conditioning/table.sql:) ENGINE=MyISAM;
contrib/forms/review_of_systems/table.sql:) ENGINE=MyISAM;
contrib/forms/review_of_systems/table.sql:) ENGINE=MyISAM;
contrib/forms/psychiatrySet/brief_aan_verwijzer/table.sql:) ENGINE=MyISAM AUTO_INCREMENT=2 ;
contrib/forms/psychiatrySet/psychiatrisch_onderzoek/table.sql:) ENGINE=MyISAM AUTO_INCREMENT=2 ;
contrib/forms/psychiatrySet/intakeverslag/table.sql:) ENGINE=MyISAM AUTO_INCREMENT=2 ;
contrib/forms/psychiatrySet/intakeverslag/autosave.sql:) ENGINE=MyISAM;
contrib/forms/cricket_injury_audit/table.sql:) ENGINE=MyISAM;
contrib/forms/evaluation/table.sql:) ENGINE=MyISAM;
contrib/forms/evaluation/table.sql:) ENGINE=MyISAM;
contrib/forms/formmaker/formscript.pl:) ENGINE=MyISAM;
contrib/forms/individual_treatment_plan/table.sql:) ENGINE=MyISAM;
contrib/forms/prior_auth/table.sql:) ENGINE=MyISAM;
contrib/forms/specialist_notes/table.sql:) ENGINE=MyISAM;
contrib/forms/treatment_protocols/table.sql:) ENGINE=MyISAM;
contrib/forms/obstetrical/table.sql:) ENGINE=MyISAM;
contrib/forms/ped_fever/table.sql:) ENGINE=MyISAM;
contrib/forms/pain/table.sql:) ENGINE=MyISAM;
contrib/forms/example2/table.sql:) ENGINE=MyISAM;
contrib/forms/example/SQL.php:	) ENGINE=MyISAM
contrib/forms/ped_GI/table.sql:) ENGINE=MyISAM;
contrib/forms/hearingtest/table.sql:) ENGINE=MyISAM;

Comment (bradymiller) [inline interface/forms/vitalsM/table.sql]: Also need to convert to InnoDB in the library/standard_tables_capture.inc script.

**Comment by olegsv** [comment]
@bradymiller added the contrib/ tables.

File: contrib/util/language_translations/buildLanguageDatabase.pl [hunk 2]
```diff
@@ -553,7 +553,7 @@ ()
   `definition` mediumtext,
   UNIQUE KEY `def_id` (`def_id`),
   KEY `cons_id` (`cons_id`) 
-) ENGINE=MyISAM AUTO_INCREMENT=".$tempCounter." ;
+) ENGINE=InnoDB AUTO_INCREMENT=".$tempCounter." ;
 \n
```
Comment (bradymiller) [inline contrib/util/language_translations/buildLanguageDatabase.pl]: oh, forgot about the language tables. To make this persistent, need to change it in the script that builds the language tables:
https://github.com/openemr/openemr/blob/master/contrib/util/language_translations/buildLanguageDatabase.pl
lines 422, 481,  556

**Comment by bradymiller** [comment]
Hi,

Regarding yehster's atomic suggestion, do you plan to try this? Note that something that also always complicates things is the logging engine in OpenEMR. Every single sql query(by default all non-select queries) are logged and inserted into the log sql table. So, if possible, a 1 query method would seem best, if possible.

-brady

**Comment by olegsv** [comment]
Hi @bradymiller . we have tested @yehster 's solution and our solution under heavy load. We started to receive deadlocks errors.
We developed  a solution to emulate current functionality under load. The solution does include four sequence tables and an ONDELETE trigger. We'll commit the changes today to you for comment.

The logging engine was the least of our problems!

**Comment by bradymiller** [comment]
Hi @olegsv ,
We have successfully avoided triggers up to now (triggers would not allow support of other non-mysql/mariadb database engines in the future); please try to solve it without a trigger :)
-brady

**Comment by olegsv** [comment]
@bradymiller what's wrong with a trigger or stored procedure?

**Comment by olegsv** [comment]
And what exactly database engines do you have in mind? :)

Today we're already using MySQL-specific features like AUTO_INCREMENT and ENUM type all across the codebase. Porting these to different DB engine would require a lot of efforts and testing.

On the contrary, MSSQL, Oracle, DB/2, Postgresql - they all have triggers.
Even if a syntax differs, you can build triggers with equivalent functionality for different databases. 
We will need to maintain  different versions of database.sql for each database engine anyway :)

Triggers, stored procedures, UDFs - they are all our friends.

**Comment by tmccormi** [comment]
I agree with Oleg, while it would be "nice" to be able to support "other" dbs, we will NEVER really do that.  we should allow the use of triggers.  The work Kevin did for Peace Corps relied on it heavily.

**Comment by sunsetsystems** [comment]
I also vaguely remember stored procedures (and I assume triggers) complicating permissions for backup/restore.  In any case I would prefer a different solution.  Atomicity doesn't matter for data integrity because PHP is not multi-threaded and probably never will be.

**Comment by bradymiller** [comment]
Hi,

We seem to go through this same discussion with triggers every couple years. Check out this thread:
https://sourceforge.net/p/openemr/discussion/202506/thread/e7e4a6ed/

Do you know of other mainstream LAMP projects that contains them on a default install?

-brady

**Comment by bradymiller** [comment]
Hi,

It also appears a number of hosts offer limited support(or none) for mysql triggers in many cases. Is it worth it to add something like this that will limit users where they can use OpenEMR?

For example:
http://www.hostinger.co.uk/knowledge-base/214

-brady

**Comment by yehster** [comment]
My suggestion when we were discussing this on Sourceforge a month ago was that Oleg and his team could use triggers for their own purposes in their own branch, because of the objections that were going to come up over using them in the official branch.

Atomicity of transactions matters even in single threaded/non-concurrent applications. That is why PHP provides support for them in the database libraries.  If it didn't matter why would mysqli and ADODB bother with the function calls for begin, commit and rollback?



https://sourceforge.net/p/openemr/discussion/202506/thread/2244a470/?page=1

> Note, I'm not proposing we use triggers for the core project. It's a possible solution for matrix to be able to use InnoDB with the incompatible tables with minimal modification to PHP code.

**Comment by olegsv** [comment]
@bradymiller no I am not familiar with any open source package that installs triggers in its schema. Neither I was  familiar with any package that has an AUTO_INCREMENT field in the end of primary keys. :) 

The trigger's logic, as we implemented it, is to reset sequence generator if all relevant records were deleted. We wanted to implement this as atomic as possible to prevent unfortunate race condition.

We can move triggers' logic inside PHP. We'll need some time to identify all places where we DELETE from the 4 affected tables.

**Comment by olegsv** [comment]
@yehster InnoDB migration project is something that all sides can benefit from.

**Comment by yehster** [comment]
There may not actually be any PHP code which deletes records from the 4 tables.

Thanks for the clarification that the trigger is only for delete.

**Comment by olegsv** [comment]
The triggers are here: https://github.com/openemr/openemr/pull/124/files  .

@yehster @sunsetsystems What would you suggest to handle  the fields that are `TEXT NOT NULL DEFAULT '' `?

**Comment by yehster** [comment]
Not sure about TEXT DEFAULT etc...
However, I wonder if using a temporary table to generate the next sequence number might work instead of needing the permanent sequence tables.  That might avoid the need for the delete triggers.

**Comment by olegsv** [comment]
@yehster how's the temporary table can hold its values between  restarts?

**Comment by yehster** [comment]
No, a temp table won't hold values between restarts  It might not be the right approach.  

I guess I don't understand why there are deadlocks when trying to generate the sequence_no without the sequence table.  My thought with the temp table was that by making the temp table the shared resource that each transaction waited on deadlocks might be avoided.  

The first time a seq_no needs to be generated in a session, the "main" table would need to be queried with ifnull max/etc... but for subsequent operations it could be generated from the temp table.

The other thing is, it might not be a big deal if the sequence numbers don't "reset" when data in the main table is deleted.  I actually don't think the behavior of autoincrement in MySQL is to reuse lower numbers after deletion.

**Comment by yehster** [comment]
http://stackoverflow.com/questions/2214141/auto-increment-after-delete-in-mysql

**Comment by sunsetsystems** [comment]
You can't use a SELECT subquery with INSERT with the same table in eadh, but try INSERT...SELECT with aggregation like this:

`SqlStatement("INSERT INTO ar_activity (pid, encounter, sequence_no, code_type, ...) " .
  "SELECT pid, encounter, COALESCE(MAX(sequence_no), 0) + 1, ?, ... " .
  "FROM ar_activity WHERE pid = ? AND encounter = ? GROUP BY (pid, encounter)",
  array(..., $pid, $encounter));`

**Comment by olegsv** [comment]
@yehster about auto_increment counter reset: if auto_increment field is a primary key, DELETE does not reset the counter, only TRUNCATE does. But for these 4 tables AFAICS we have different scenario where delete for Auto-increment field really does restart the counter. consider the following easy test:

`
set sql_mode="";
create table testkeys( field1 int not null, field2 int not null, field3 int not null auto_increment, primary key ( field1, field2, field3 ) ) engine=MyISAM;
insert into testkeys(field1,field2) values(1,1),(1,1),(1,1),(2,2),(2,2),(2,2),(3,3),(3,3),(3,3);
select * from testkeys order by field1, field2;
-- AI seq restarts for each combination of field1 and field2
delete from testkeys where field1=1 and field2=1;
-- will it restart the value of field3?
insert into testkeys(field1,field2) values(1,1),(1,1),(1,1);
select * from testkeys order by field1, field2;
-- Yes, it will.
`

**Comment by olegsv** [comment]
@yehster @sunsetsystems about the deadlocks. We have prepared a small test to test race condition.
The idea of the script is to run this ` INSERT INTO ar_activity .... SELECT IFNULL(MAX(sequence_id))+1 AS sequence_no ` via mysqli driver, repeat it with same values for pid and encounter 5000 times, and exit on mysql error. Then we run 4-5 instances simultaneously.

**Comment by yehster** [comment]
It's unlikely that there would every be that much concurrent activity in these tables, based on what they represent.

It might be appropriate to make a trade-off between "no-triggers" and "maximum potential performance".

**Comment by sunsetsystems** [comment]
I can't imagine my solution above producing deadlocks as it's a single query.

**Comment by olegsv** [comment]
@sunsetsystems take a look: [http://pastebin.com/XLfmzG8w](http://pastebin.com/XLfmzG8w)

**Comment by sunsetsystems** [comment]
@olegsv This merits a few comments.
1. Your point is taken that a single statement can produce a 1213 error, however that's not really an "error", it's just mysql asking you to wait a few milliseconds and try again.
2. This applies to command line PHP where real concurrency can happen, however that's not the case with a PHP web script.
3. The statement you tested is not the one I proposed. I'm surprised you did not get a syntax error, as MySQL's documentation of subqueries says "you cannot modify a table and select from the same table in a subquery". Please read my post again carefully.
4. Thanks for making me do some reading about ACID with InnoDB.  :)

**Comment by olegsv** [comment]
@sunsetsystems re: 1. How you are supposed to process this message from MySQL ? 

2. Command line is just a way to simluate  race condition under high load. Technically there's no difference from the web.

We'll do testing with your query and see how it goes.

**Comment by olegsv** [comment]
@sunsetsystems  is this SQL query what you proposed?

INSERT INTO `openemr`.`ar_activity`
  ( `pid`, `encounter`,
  `sequence_no`,
 `code_type`,`code`,`payer_type`,`post_time`,`post_user`,`session_id`,`modified_time`,`follow_up`,`follow_up_note`,`account_code` )

SELECT 1,1, 
    COALESCE( MAX(sequence_no), 0)+1 ,
    "type1","code1",1,now(),1,1,now(),'Y','',"acct1"
    FROM ar_activity WHERE pid=1 AND encounter=1 GROUP BY pid, encounter

**Comment by sunsetsystems** [comment]
If you use CLI then I think you need to also test for error 1213, and when it occurs pause for a few msec and try the query again.  I understand that is just an oddity of how MySQL works.

I'm pretty sure CLI is different from web with respect to PHP concurrency.

Yes that query looks like what I mean.

**Comment by sunsetsystems** [comment]
By the way try the usleep() function for the pause operation.  But I would not do that for a web script.

**Comment by sunsetsystems** [comment]
To expand a bit on CLI vs. web... PHP with Apache depends on apache2-mpm-prefork which is the non-threaded Apache model, because mod_php is not thread-safe.

**Comment by olegsv** [comment]
Then try your query on an empty ar_activity table.

BTW using sleep gives the same results, just after a few more seconds.

**Comment by sunsetsystems** [comment]
What results are those?  What error?

**Comment by olegsv** [comment]
Same error 1213... Just a few seconds later. Have you tried to run the query?

**Comment by sunsetsystems** [comment]
No I have not tried it.  How can you be getting error 1213 if you are pausing and re-trying when getting that?  If you are just giving up after a few pauses then I think you are expecting too much.  You have several processes continually fighting over the same record lock so there will be a LOT of conflicts.

**Comment by olegsv** [comment]
The script was a demo to simulate an unfortunate event. How would  you change the application's code,

**Comment by olegsv** [comment]
And please try to run the query on an empty table.

**Comment by sunsetsystems** [comment]
Try the attached.  Empty table won't work here, you have to pre-seed with one entry and the real code will need to do something about that.

[test_insert.php.txt](https://github.com/openemr/openemr/files/276546/test_insert.php.txt)

**Comment by sunsetsystems** [comment]
Here's a version that works with no existing rows.

[test_insert.php.txt](https://github.com/openemr/openemr/files/276569/test_insert.php.txt)

**Comment by olegsv** [comment]
@sunsetsystems Note you created internal loop. With this loop, our SQL will work too, without problems and without the need of additional tables.


@bradymiller How do we continue from here?

**Comment by sunsetsystems** [comment]
@olegsv The problem with your subquery is that mysql docs say it is not allowed.  Another problem with both yours and mine is they do not return the generated sequence_no, so in spite of all this discussion it may be better (and more understandable) to split it into two queries (SELECT and then INSERT), and wrap them in a transaction.

To proceed, suggest you write common functions for inserting into the tables in question and call those where they apply.

**Comment by olegsv** [comment]
Another possibility is to split the fix into two. let's leave these 4 problematic tables aside and fix them later.

BTW Can you refer to mysql page that's saying so? Because AFAIK this is perfectly legitimate query, as far as SELECT returning 1 row. Found this as far as in 2012: 
http://stackoverflow.com/questions/617493/can-i-use-a-subquery-inside-an-insert-statement

Can you refer me to place in  the code where I need to return the sequence_no from an insert?

**Comment by sunsetsystems** [comment]
Sure, go here and search for "same table":
https://dev.mysql.com/doc/refman/5.7/en/subqueries.html

I don't know of a place that requires the sequence_no from the insert.  Looked around a bit but didn't find one.  But it might apply to some of the other tables.

**Comment by olegsv** [comment]
In C.4 about restrictions they have a bit different example, where subquery's SELECT replaces all the values. In our query, only a value is loaded from (SELECT) and is stored temporary table.

**Comment by yehster** [comment]
This "exception" might be what is allowing Oleg's version of the query to work.


https://dev.mysql.com/doc/mysql-reslimits-excerpt/5.7/en/subquery-restrictions.html

Exception: The preceding prohibition does not apply if you are using a subquery for the modified table in the FROM clause. Example:

UPDATE t ... WHERE col = (SELECT * FROM (SELECT ... FROM t...) AS _t ...);

Here the result from the subquery in the FROM clause is stored as a temporary table, so the relevant rows in t have already been selected by the time the update to t takes place.

**Comment by sunsetsystems** [comment]
If there is assurance from the mysql docs that the subquery is safe then I have no problem with it.

File: sql/database.sql [hunk 3]
```diff
@@ -2309,7 +2309,7 @@ CREATE TABLE `icd10_pcs_order_code` (
   `revision` int default 0,
   KEY `pcs_code` (`pcs_code`),
   KEY `active` (`active`)
-) ENGINE=MyISAM;
+) ENGINE=InnoDBInnoDB;
```
Comment (bradymiller) [inline sql/database.sql]: typo

File: sql/database.sql [hunk 4]
```diff
@@ -8085,3 +8085,53 @@ CREATE TABLE `form_clinical_instructions` (
...
+-- --------------------------------------------------------
+--
+-- Table structure for table `procedure_order_code_sequences`
+--
+CREATE TABLE IF NOT EXISTS `procedure_order_code_seq` (
+  `procedure_order_id` bigint(20) NOT NULL,
+  `counter` int(11) unsigned NOT NULL,
+  PRIMARY KEY (`procedure_order_id`)
+) ENGINE=InnoDB;
+-- --------------------------------------------------------
```
Comment (bradymiller) [inline sql/database.sql]: place these tables within above file in alphabetical order and mimick formatting of the above tables.
ie. DROP TABLE IF EXISTS etc.

File: library/sql.inc [hunk 5]
```diff
@@ -650,4 +650,151 @@ function generic_sql_insert_id()
...
+
+        case 'claims':
+            //remove claims sequences
+            $rows_deleted = sqlStatement("SELECT s.patient_id,s.encounter_id
+                                FROM `claims_seq` s LEFT JOIN
+                                claims s1 ON s1.patient_id=s.patient_id and s1.encounter_id=s.encounter_id
+                                WHERE s1.patient_id IS NULL");
+
+            while ($row = sqlFetchArray($rows_deleted)) {
+                sqlStatement("DELETE FROM claims_seq WHERE patient_id = {$row['patient_id']} AND encounter_id = {$row['encounter_id']}");
```
Comment (bradymiller) [inline library/sql.inc]: use binding

File: library/sql.inc [hunk 6]
```diff
@@ -650,4 +650,151 @@ function generic_sql_insert_id()
...
+
+        case 'ar_activity':
+            //remove ar_activity sequences
+            $rows_deleted = sqlStatement("SELECT s.pid,s.encounter
+                        FROM `ar_activity_seq` s
+                        LEFT JOIN ar_activity s1 ON s1.pid=s.pid and s1.encounter=s.encounter
+                        WHERE s1.pid IS NULL");
+
+            while ($row = sqlFetchArray($rows_deleted)) {
+                    sqlStatement("DELETE FROM ar_activity_seq WHERE pid = {$row['pid']} AND encounter = {$row['encounter']}");
```
Comment (bradymiller) [inline library/sql.inc]: use binding

File: library/sql.inc [hunk 7]
```diff
@@ -650,4 +650,151 @@ function generic_sql_insert_id()
...
+
+        case 'procedure_answers':
+            $rows_deleted = sqlStatement("SELECT s.procedure_order_id,s.procedure_order_seq, s.question_code
+                                      FROM `procedure_answers_seq` s LEFT JOIN
+                                      procedure_answers s1 ON s1.procedure_order_id=s.procedure_order_id AND s1.procedure_order_seq=s.procedure_order_seq AND s1.question_code = s.question_code
+                                      WHERE s1.procedure_order_id IS NULL
+                                      ");
+
+            while ($row = sqlFetchArray($rows_deleted)) {
+                    sqlStatement("DELETE FROM procedure_answers_seq WHERE procedure_order_id = {$row['procedure_order_id']} AND procedure_order_seq = {$row['procedure_order_seq']}  AND question_code = {$row['question_code']}");
```
Comment (bradymiller) [inline library/sql.inc]: use binding

File: library/sql.inc [hunk 8]
```diff
@@ -650,4 +650,151 @@ function generic_sql_insert_id()
...
+            }
+            break;
+
+        case 'procedure_order_code':
+            $rows_deleted = sqlStatement("SELECT s.procedure_order_id
+                                      FROM `procedure_order_code_seq` s LEFT JOIN
+                                      procedure_order_code s1 ON s1.procedure_order_id=s.procedure_order_id
+                                      WHERE s1.procedure_order_id IS NULL");
+            while ($row = sqlFetchArray($rows_deleted)) {
+                    sqlStatement("DELETE FROM procedure_order_code_seq WHERE procedure_order_id = {$row['procedure_order_id']}");
```
Comment (bradymiller) [inline library/sql.inc]: use binding

**Comment by bradymiller** [comment]
wow, this is some serious coding...
Is the current code the planned solution? If so, will begin testing it out(especially upgrading) and will point a up for grabs demo to it.
And just to clarify, what ended up being the plan for TEXT fields default setting to not error out in INNODB?

**Comment by olegsv** [comment]
This looks more or less final.
I have written a few words here: http://www.open-emr.org/wiki/index.php/InnoDB_Transition

File: interface/billing/edit_payment.php [hunk 9]
```diff
@@ -176,9 +176,10 @@
 				 }
 				else
 				 {
-				  sqlStatement("insert into ar_activity set "    .
+					 sqlStatement("insert into ar_activity set "    .
 					"pid = '"       . trim(formData("HiddenPId$CountRow"   )) .
```
Comment (sunsetsystems) [inline interface/billing/edit_payment.php]: Please create a common callable function for inserting into ar_activity, and call that here and everywhere else you insert into it.

File: interface/billing/edit_payment.php [hunk 10]
```diff
@@ -176,9 +176,10 @@
...
 					"', encounter = '"     . trim(formData("HiddenEncounter$CountRow"   ))  .
+					"', sequence_no = '"     . getCurrentSequence('ar_activity', array('pid' => trim(formData("HiddenPId$CountRow"   )), 'encounter' => trim(formData("HiddenEncounter$CountRow"   )))) .
```
Comment (sunsetsystems) [inline interface/billing/edit_payment.php]: We decided we don't need a new sequence number table, right?

File: interface/forms/procedure_order/new.php [hunk 11]
```diff
@@ -113,20 +113,23 @@ function getListOptions($list_id , $fieldnames=array('option_id', 'title', 'seq'
...
+  removeSequence('procedure_answers');
+  removeSequence('procedure_order_code');
 
   for ($i = 0; isset($_POST['form_proc_type'][$i]); ++$i) {
     $ptid = $_POST['form_proc_type'][$i] + 0;
     if ($ptid <= 0) continue;
 
     $prefix = "ans$i" . "_";
 
-    $poseq = sqlInsert("INSERT INTO procedure_order_code SET ".
```
Comment (sunsetsystems) [inline interface/forms/procedure_order/new.php]: Same comments as with ar_activity.  No sequence number table, right?

File: library/billing.inc [hunk 12]
```diff
@@ -174,21 +174,20 @@ function updateClaim($newversion, $patient_id, $encounter_id, $payer_id=-1, $pay
...
-     {//Claim automatic forward case.
-     
-     $sql = "INSERT INTO claims SET " .
-      "patient_id = '$patient_id', " .
-      "encounter_id = '$encounter_id', " .
-      "bill_time = NOW(), status=$status";
-     
-     }
-    
+      $sql .= "INSERT INTO claims SET " .
```
Comment (sunsetsystems) [inline library/billing.inc]: Same comments as for ar_activity.

File: library/sql.inc [hunk 13]
```diff
@@ -650,4 +650,151 @@ function generic_sql_insert_id()
...
+ * This feature create internal auto increment for every multiple-column index in the table.
+ * More info - http://dev.mysql.com/doc/refman/5.6/en/example-auto-increment.html#idm139866976916160
+ * For imitate it the system manage custom sequences in a few tables.
+ * @param (string) table name
+ * @param (array) primary keys (without auto increment column). format - array(column_name => value)
+ * @return (int) current sequence number
+ */
+function getCurrentSequence($table, array $primaryKeys){
+
+    $sequencesTable = $table . '_seq';
```
Comment (sunsetsystems) [inline library/sql.inc]: I assume this function is no longer needed, given recent discussions.

**Comment by sunsetsystems** [comment]
Didn't we decide we don't need the sequence tables?

Also there should be a shared function for inserts into ar_activity and perhaps also the other tables.

**Comment by olegsv** [comment]
I can see why we need to refactor inserts into a function. Perhaps doing this later would be better?

Regarding sequence tables. With ondelete triggers removed,  I agree the solution is still clumsier. However, would you agree that this scheme provides more protection against unfortunate race condition?

**Comment by sunsetsystems** [comment]
Later is fine for converting the 4 tables in question to InnoDB.

Not sure which method "this scheme" is, but I will assume that mysql transaction support is not broken.  So if the called function needs to return what used to be the auto_increment value, then I suggest doing SELECT then INSERT separately inside a transaction.  Otherwise one of the single queries as discussed here is OK.

**Comment by bradymiller** [comment]
Hi,

So, then we have 2 possible solutions for the "fugitive" tables as detailed here:
http://www.open-emr.org/wiki/index.php/InnoDB_Transition

Would it make more sense to go with solution 2 for now since it seems less invasive on the code? Or does solution 1 produce less chance of deadlock?

I do think there is a light at the end of this tunnel and my vote is to get to a full INNODB conversion (rather than delay the 4 "fugitive" tables) in since I think this opens the door to users being able to use other third party mysql tools. My guess is that there will never be as much effort on this as there is at this point in time (4 developers actively involved in the solution).

-brady

**Comment by olegsv** [comment]
Solution 1 does not produce deadlock on insert. It requires additional tables and additional logic. We did not test it for race condition upon DELETE.

Solution 2 may produce a deadlock error 1213 if two requests with same values of pid and encounter will be submitted in the same time.

**Comment by sunsetsystems** [comment]
Again, solution 2 can produce an error 1213 which is MySQL's way of telling the caller to try again.  As we have already seen in this thread, that can be handled and does not even come up in single-threaded web scripts which is the case when using mod_php under Apache.

My suggestion is to let Oleg wrap up his commit *excluding* the "fugitive" tables, and I will follow with a commit to convert those to InnoDB.

**Comment by olegsv** [comment]
We had already developed  complete solution  without the use of sequence tables , i.e. Solution 2.
We will prepare and send the commit including all the tables later today. 
From there @sunsetsystems will bring this to perfection.

**Comment by bradymiller** [comment]
Sounds reasonable. This has turned into a rather messy branch. Let me know if you need guidance on flattening this code into 1 commit.

**Comment by olegsv** [comment]
We have opened pull request 129 to replace this.

**Pull request closed without merging**